### PR TITLE
fix #1213: upgrade to latest LSP

### LINKED ISSUE #1213: Update to a more recent vscode-languageclient
It would be useful for me if we could update to a more recent vscode-languageclient.  In particular, I would like if we used a version that supports the [MarkupContent](https://microsoft.github.io/language-server-protocol/specification#new-markupcontent) response to the Hover request.  Currently, our client only understand MarkupString, but that one is deprecated.  Since I'm adding support for Hover in clangd, I'd like to use the non-deprecated response.

According to @akosyakov, this issue

  https://github.com/TypeFox/monaco-languageclient/issues/24

needs to be resolved before we can upgrade.

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by AlexTugarev** [comment]
```
Uncaught Error: marked(): input parameter is undefined or null

Error: marked(): input parameter is undefined or null
    at Object.marked (:3000/vs/editor/editor.main.js:5476)
    at Object.renderMarkdown (:3000/vs/editor/editor.main.js:17218)
    at MarkdownRenderer.render (:3000/vs/editor/editor.main.js:71884)
    at SuggestionDetails.render (:3000/vs/editor/editor.main.js:101836)
    at SuggestWidget.showDetails (:3000/vs/editor/editor.main.js:102364)
    at :3000/vs/editor/editor.main.js:102091
    at Object.notifySuccess [as _notify] (:3000/vs/editor/editor.main.js:11025)
    at Object.enter (:3000/vs/editor/editor.main.js:10707)
    at Promise_ctor._run (:3000/vs/editor/editor.main.js:10923)
    at Promise_ctor._completed (:3000/vs/editor/editor.main.js:10891)
    at Object.marked (:3000/vs/editor/editor.main.js:5476)
    at Object.renderMarkdown (:3000/vs/editor/editor.main.js:17218)
    at MarkdownRenderer.render (:3000/vs/editor/editor.main.js:71884)
    at SuggestionDetails.render (:3000/vs/editor/editor.main.js:101836)
    at SuggestWidget.showDetails (:3000/vs/editor/editor.main.js:102364)
    at :3000/vs/editor/editor.main.js:102091
    at Object.notifySuccess [as _notify] (:3000/vs/editor/editor.main.js:11025)
    at Object.enter (:3000/vs/editor/editor.main.js:10707)
    at Promise_ctor._run (:3000/vs/editor/editor.main.js:10923)
    at Promise_ctor._completed (:3000/vs/editor/editor.main.js:10891)
    at :3000/vs/editor/editor.main.js:12476
```
I see this error with content assist for Java.

**Comment by AlexTugarev** [comment]
```
[Error - 6:50:51 PM] Jul 26, 2018 6:50:32 PM End Of File
End Of File
org.eclipse.core.runtime.CoreException: End Of File
at org.eclipse.jdt.internal.corext.dom.TokenScanner.readNext(TokenScanner.java:150)
at org.eclipse.jdt.internal.corext.dom.TokenScanner.readNext(TokenScanner.java:188)
at org.eclipse.jdt.ls.core.internal.contentassist.SnippetCompletionProposal.needsPublic(SnippetCompletionProposal.java:139)
at org.eclipse.jdt.ls.core.internal.contentassist.SnippetCompletionProposal.getSnippets(SnippetCompletionProposal.java:78)
at org.eclipse.jdt.ls.core.internal.handlers.CompletionHandler.computeContentAssist(CompletionHandler.java:111)
at org.eclipse.jdt.ls.core.internal.handlers.CompletionHandler.completion(CompletionHandler.java:41)
at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$4(JDTLanguageServer.java:425)
at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$22(JDTLanguageServer.java:716)
at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

JDT LS is crying when empty Java file is opened.

**Comment by AlexTugarev** [comment]
```
Error: unsupported
    at Object.unsupported [as waitUntil] (vscode-api.ts:15)
    at WillSaveWaitUntilFeature.callback (client.js:527)
    at onWillSaveTextDocument (vscode-api.ts:146)
    at CallbackList.../../node_modules/vscode-jsonrpc/lib/events.js.CallbackList.invoke (events.js:71)
    at Emitter.../../node_modules/vscode-jsonrpc/lib/events.js.Emitter.fire (events.js:135)
    at resolve (monaco-workspace.ts:156)
    at new Promise (<anonymous>)
    at MonacoWorkspace.fireWillSave (monaco-workspace.ts:155)
    at model.onWillSaveModel.event (monaco-workspace.ts:133)
    at CallbackList.invoke (event.ts:107)
log @ logger.ts:107
../../node_modules/vscode-jsonrpc/lib/events.js.CallbackList.invoke @ events.js:74
../../node_modules/vscode-jsonrpc/lib/events.js.Emitter.fire @ events.js:135
resolve @ monaco-workspace.ts:156
fireWillSave @ monaco-workspace.ts:155
model.onWillSaveModel.event @ monaco-workspace.ts:133
invoke @ event.ts:107
fire @ event.ts:199
resolve @ monaco-editor-model.js:261
fireWillSaveModel @ monaco-editor-model.js:260
(anonymous) @ monaco-editor-model.js:241
(anonymous) @ monaco-editor-model.js:22
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:18
doSave @ monaco-editor-model.js:237
run @ monaco-editor-model.js:193
(anonymous) @ monaco-editor-model.js:131
(anonymous) @ monaco-editor-model.js:22
push.../../packages/monaco/lib/browser/monaco-editor-model.js.__awaiter @ monaco-editor-model.js:18
pendingOperation.pendingOperation.then @ monaco-editor-model.js:129
Promise.then (async)
run @ monaco-editor-model.js:129
scheduleSave @ monaco-editor-model.js:193
save @ monaco-editor-model.js:126
(anonymous) @ saveable.ts:63
(anonymous) @ saveable.ts:15
../../packages/core/lib/browser/saveable.js.__awaiter @ saveable.ts:15
save @ saveable.ts:60
(anonymous) @ application-shell.ts:1231
(anonymous) @ application-shell.ts:15
../../packages/core/lib/browser/shell/application-shell.js.__awaiter @ application-shell.ts:15
save @ application-shell.ts:1230
execute @ common-frontend-contribution.ts:380
tryKeybindingExecution @ keybinding.ts:424
run @ keybinding.ts:458
document.addEventListener.event @ frontend-application.ts:147
18:52:30.752 messages.js:46 Uncaught (in promise) Error: Unhandled method toJSON
    at new ResponseError (messages.js:46)
    at handleResponse (main.js:430)
    at processMessageQueue (main.js:258)
    at main.js:242
    at run (setImmediate.js:40)
    at runIfPresent (setImmediate.js:69)
    at onGlobalMessage (setImmediate.js:109)
```

on save of a Java file.

File: packages/languages/src/browser/window-impl.ts [hunk 1]
```diff
@@ -56,8 +56,17 @@ export class WindowImpl implements Window {
...
-            show: function (preserveFocus?: boolean): void {
-                // no-op
+            show: async (preserveFocus?: boolean) => {
+                outputChannel.setVisibility(true);
+                const options = Object.assign({
+                    preserveFocus: false,
+                }, { preserveFocus });
+                const activate = !options.preserveFocus;
+                const reveal = options.preserveFocus;
+                this.outputContribution.openView({ activate, reveal });
```
Comment (AlexTugarev) [inline packages/languages/src/browser/window-impl.ts]: really reveal? if it's already opened but hidden behind another widget, I personally wouldn't like it to bother me. e.g. I'd prefer to see the watcher output in terminal.

**Comment by AlexTugarev** [comment]
```
errors.ts:81 Uncaught Error: Cannot read property 'isDisposed' of undefined

TypeError: Cannot read property 'isDisposed' of undefined
    at Function.getOptionsFromEditor (text-editor-main.ts:364)
    at Function.readFromEditor (text-editor-main.ts:342)
    at TextEditorMain.updateProperties (text-editor-main.ts:56)
    at editorListeners.push.monaco.onDidScrollChange (text-editor-main.ts:94)
    at Emitter.fire (event.ts:140)
    at ViewOutgoingEvents.viewEventBus.onDidScroll (codeEditorWidget.ts:433)
    at ViewOutgoingEvents.emitScrollChanged (viewOutgoingEvents.ts:45)
    at View.onScrollChanged (viewImpl.ts:333)
    at ViewEventHandler.handleEvents (viewEventHandler.ts:164)
    at ViewEventDispatcher._doConsumeQueue (viewEventDispatcher.ts:89)
    at Function.getOptionsFromEditor (text-editor-main.ts:364)
    at Function.readFromEditor (text-editor-main.ts:342)
    at TextEditorMain.updateProperties (text-editor-main.ts:56)
    at editorListeners.push.monaco.onDidScrollChange (text-editor-main.ts:94)
    at Emitter.fire (event.ts:140)
    at ViewOutgoingEvents.viewEventBus.onDidScroll (codeEditorWidget.ts:433)
    at ViewOutgoingEvents.emitScrollChanged (viewOutgoingEvents.ts:45)
    at View.onScrollChanged (viewImpl.ts:333)
    at ViewEventHandler.handleEvents (viewEventHandler.ts:164)
    at ViewEventDispatcher._doConsumeQueue (viewEventDispatcher.ts:89)
    at errors.ts:81
```
might be related to new monaco version?

File: configs/base.tsconfig.json [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
+    "noEmitOnError": false,
+    "noImplicitThis": true,
+    "noUnusedLocals": true,
+    "strictNullChecks": true,
+    "experimentalDecorators": true,
+    "emitDecoratorMetadata": true,
+    "downlevelIteration": true,
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "target": "es6",
```
Comment (AlexTugarev) [inline configs/base.tsconfig.json]: -1 unfortunately!

File: packages/callhierarchy/src/browser/utils.ts [hunk 3]
```diff
@@ -48,7 +48,10 @@ export function filterSame(locations: Location[], definition: Location): Locatio
     );
 }
 
-export function filterUnique(locations: Location[]): Location[] {
+export function filterUnique(locations: Location[] | null): Location[] {
+    if (!locations) {
```
Comment (AlexTugarev) [inline packages/callhierarchy/src/browser/utils.ts]: nice!

File: packages/languages/src/browser/window-impl.ts [hunk 4]
```diff
@@ -56,8 +56,17 @@ export class WindowImpl implements Window {
         return {
             append: outputChannel.append.bind(outputChannel),
             appendLine: outputChannel.appendLine.bind(outputChannel),
-            show: function (preserveFocus?: boolean): void {
-                // no-op
+            show: async (preserveFocus?: boolean) => {
```
Comment (AlexTugarev) [inline packages/languages/src/browser/window-impl.ts]: thanks! very cool!

one thing: for Java files it doesn't select the source correctly, stays on TypeScript.

**Comment by AlexTugarev** [comment]
@akosyakov, have you update entire yarn.lock? if so, we need to check more then just editors and LS'.

File: packages/languages/src/browser/window-impl.ts (hunk 1 above)
Comment (akosyakov) [reply packages/languages/src/browser/window-impl.ts]: It depends how the language client is configured, by default it won't be displayed.

File: packages/languages/src/browser/window-impl.ts (hunk 4 above)
Comment (AlexTugarev) [reply packages/languages/src/browser/window-impl.ts]: e.g. create and open an empty java file. JDT LS will throw, you should see it there.

**Comment by akosyakov** [comment]
if the build is green, I'm merging it.

**Pull request merged**